### PR TITLE
fix: check VM/VMSS state before PATCH calls

### LINKED ISSUE #926: Check VMSS state before PATCH calls
**Describe the request**

**Explain why AAD Pod Identity needs it**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by chewong**
Signed-off-by: Ernest Wong <chuwon@microsoft.com>

**Reason for Change**:

Fixes #926.

**Requirements**

- [x] squashed commits
- [x] included documentation
- [x] added unit tests and e2e tests (if applicable). See [test standard](https://github.com/Azure/aad-pod-identity/blob/master/CONTRIBUTING.md#test-standard) for more details.
- [x] ran `make precommit`

**Issue Fixed**:

**Please answer the following questions with yes/no**:

Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

- [ ] yes
- [x] no

**Notes for Reviewers**:

### CODE REVIEW

File: pkg/cloudprovider/vm.go [hunk 1]
```diff
@@ -99,6 +99,10 @@ func (c *VMClient) Get(rgName string, nodeName string) (compute.VirtualMachine,
 
 // UpdateIdentities updates the user assigned identities for the provided node
 func (c *VMClient) UpdateIdentities(rg, nodeName string, vm compute.VirtualMachine) error {
+	if *vm.ProvisioningState != string(compute.ProvisioningStateSucceeded) {
```
Comment (chewong) [inline pkg/cloudprovider/vm.go:102]: Let me know if there are other states that we want to check

Comment (aramase) [reply pkg/cloudprovider/vm.go:102]: I think just waiting for `ProvisioningStateSucceeded` might cause delay in the identity assignment operations. We should check if the `ProvisioningState` is `Deleting` and skip operations only for those.

File: pkg/cloudprovider/vm.go [hunk 2]
```diff
@@ -99,6 +99,10 @@ func (c *VMClient) Get(rgName string, nodeName string) (compute.VirtualMachine,
 
 // UpdateIdentities updates the user assigned identities for the provided node
 func (c *VMClient) UpdateIdentities(rg, nodeName string, vm compute.VirtualMachine) error {
+	if *vm.ProvisioningState == string(compute.ProvisioningState1Deleting) {
```
Comment (aramase) [inline pkg/cloudprovider/vm.go:102]: Why is this `ProvisioningState1Deleting` instead of `ProvisioningStateDeleting`?

Comment (chewong) [reply pkg/cloudprovider/vm.go:102]: fixed

File: pkg/cloudprovider/vmss.go [hunk 3]
```diff
@@ -136,7 +140,7 @@ func (c *VMSSClient) Get(rgName string, vmssName string) (ret compute.VirtualMac
 		resp := vmss.Response.Response
 		// Update RetryAfterReader so that no more requests would be sent until RetryAfter expires.
 		c.retryAfterReader = time.Now().Add(getRetryAfter(resp))
-		return vmss, fmt.Errorf("failed to get vmss %s in resource group %s, error: %+v", vmssName, rgName, err)
+		return compute.VirtualMachineScaleSet{}, fmt.Errorf("failed to get vmss %s in resource group %s, error: %+v", vmssName, rgName, err)
```
Comment (aramase) [inline pkg/cloudprovider/vmss.go:143]: nit: We should also make this change for `vm.go`

**Review by aramase — APPROVED** [review]

**Pull request merged**